### PR TITLE
BUG: Fixed Von Mises distribution for big values of kappa

### LINKED ISSUE #17275: numpy.random.vonmises() fails for kappa > ~10^8
The von Mises distribution should limit to a Gaussian distribution as kappa -> infinity. The function numpy.random.vonmises() appears to capture this behavior for large values up kappa (approximately up to 10^8). However, beyond this, the distribution starts to deviate significantly from a Gaussian. Looks more like a Cantor set.

### Reproducing code example:

```python
import numpy as np
import matplotlib.pyplot as plt

# Reasonable Case
plt.hist(np.random.vonmises(0,10**7,100000),bins=200)
plt.show()

# Broken Case –– Should closely approximate a Gaussian.
plt.hist(np.random.vonmises(0,10**15,100000),bins=200)
plt.show()
```

### PR DESCRIPTION
**Pull request by Androp0v**
Fixed Von Mises distribution for big values of kappa by falling back to a normal distribution (which the von mises distribution converges to) after kappa > 10^6 to fix #17275. 

The mean squared error between the von mises distribution and normal distribution was shown to grow after kappa > 10^6 before the fix instead of converging, due to floating point precision errors in the acos() calculation.
Figure_1

The von mises function already fell back to the uniform distribution for kappa < 10^(-8) so it seemed appropriate to fall back to the normal distribution on the upper end (kappa > 10^6) since the difference with the normal distribution is vanishingly small.

### CODE REVIEW

File: numpy/random/src/distributions/distributions.c [hunk 1]
```diff
@@ -853,9 +853,14 @@ double random_vonmises(bitgen_t *bitgen_state, double mu, double kappa) {
        */
       s = (1. / kappa + kappa);
     } else {
-      double r = 1 + sqrt(1 + 4 * kappa * kappa);
-      double rho = (r - sqrt(2 * r)) / (2 * kappa);
-      s = (1 + rho * rho) / (2 * rho);
+      /* Fallback to normal distribution for big values of kappa*/
+      if (kappa > 1e6){
+        return mu + sqrt(1/kappa) * random_standard_normal(bitgen_state);
+      }else{
```
Comment (charris) [inline numpy/random/src/distributions/distributions.c:859]: Do
```
}
else {
```
That isn't uniformly used here, but it is the NumPy standard.

File: numpy/random/src/distributions/distributions.c [hunk 2]
```diff
@@ -853,9 +853,14 @@ double random_vonmises(bitgen_t *bitgen_state, double mu, double kappa) {
        */
       s = (1. / kappa + kappa);
     } else {
-      double r = 1 + sqrt(1 + 4 * kappa * kappa);
-      double rho = (r - sqrt(2 * r)) / (2 * kappa);
-      s = (1 + rho * rho) / (2 * rho);
+      /* Fallback to normal distribution for big values of kappa*/
+      if (kappa > 1e6){
```
Comment (charris) [inline numpy/random/src/distributions/distributions.c:857]: Missing spaces before `*/` and `{`.

**Comment by charris** [comment]
Couple of white space nits. I think this is OK, although it may affect legacy also. @bashtage Thoughts?

**Comment by WarrenWeckesser** [comment]
Assuming this is a reasonable approach (I haven't checked), there is still a problem:  the values must be wrapper to the interval [-π, π].  In this example, values greater than π are generated:

```
In [41]: import numpy as np

In [42]: np.__version__
Out[42]: '1.20.0.dev0+e2f3e01'

In [43]: rng = np.random.default_rng()

In [44]: x = rng.vonmises(3.14159, 2e6, size=15)

In [45]: x
Out[45]: 
array([3.14307189, 3.14154359, 3.14150478, 3.14176629, 3.14154238,
       3.14219684, 3.14149787, 3.14055073, 3.14238474, 3.14247015,
       3.14303787, 3.14204343, 3.14188761, 3.14127442, 3.14131087])

In [46]: x > np.pi
Out[46]: 
array([ True, False, False,  True, False,  True, False, False,  True,
        True,  True,  True,  True, False, False])
```

**Comment by Androp0v** [comment]
True. I will update the PR accounting for values close to -π or π.

**Comment by bashtage** [comment]
Could you add some tests, including @WarrenWeckesser 's to make sure the bounds are respected?

**Comment by Androp0v** [comment]
Yes. Added a test for values with a mu equal to -π and π and a value of kappa that already uses the gaussian approximation (actually one of the smallest kappa values that triggers the fallback to gaussian, to have the biggest possible standard deviation in the distribution).

**Comment by charris** [comment]
@Androp0v Thanks. I've merged #18498 that builds on top of this, your commits are preserved.

**Pull request closed without merging**